### PR TITLE
[stdlib][Qol] SR-11295 Removing stdlib unnecessary coercions

### PR DESCRIPTION
**Pull request by LucianoPAlmeida**
This removes unnecessary 'as' same type casts.
Extension of #26710, just separating this in other PR. 

cc @xedin

### CODE REVIEW

File: stdlib/public/Darwin/Foundation/NSError.swift [hunk 1]
```diff
@@ -235,7 +235,7 @@ public func _getErrorDefaultUserInfo<T: Error>(_ error: T)
       _errorDomainUserInfoProviderQueue.sync {
         if NSError.userInfoValueProvider(forDomain: domain) != nil { return }
         NSError.setUserInfoValueProvider(forDomain: domain) { (nsError, key) in
-          let error = nsError as Error
+          let error = nsError
```
Comment (natecook1000) [inline stdlib/public/Darwin/Foundation/NSError.swift]: If we aren't performing this cast, we don't need a distinct `error` and `nsError` here, right? Cleanest to probably just rename the parameters in the line above.

**Review by natecook1000 — COMMENTED** [review]
Thanks, @LucianoPAlmeida! One small nit, otherwise looks good 👍

File: stdlib/public/Darwin/Foundation/NSError.swift (hunk 1 above)
Comment (LucianoPAlmeida) [reply stdlib/public/Darwin/Foundation/NSError.swift]: Right, Thank's @natecook1000 :)

**Comment by LucianoPAlmeida** [comment]
Just updating the initial description
The original diagnostic PR was reverted because of an issue with type variables and contextual types, so it is being re-applied on #27895 :)

**Comment by CodaFi** [comment]
@swift-ci please smoke test

**Comment by LucianoPAlmeida** [comment]
Thank's @CodaFi :) 
The linux failure seems unrelated

**Comment by LucianoPAlmeida** [comment]
Although the re-apply of the diagnostic is still in the workings, this is actually not dependent on it. And since it is only very simple removals, I was wondering if we get this in if everything is ok?  
Also, can someone with rights kick the CI? if everything is ok of course :)

**Comment by CodaFi** [comment]
Sure.  In the future, though, could you please pull with a rebase so there aren’t intervening merge commits?  If this passes, I’ll squash this one out for you.

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by LucianoPAlmeida** [comment]
Thank you @CodaFi 
> In the future, though, could you please pull with a rebase so there aren’t intervening merge commits? 

Sure :)

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**